admin/broadleaf-open-admin-platform/src/main/java/org/broadleafcommerce/openadmin/server/service/persistence/module/FieldManager.java

```diff
@@ -48,37 +49,57 @@
 import javax.persistence.metamodel.EntityType;
 
 /**
- * 
+ *
  * @author jfischer
  *
  */
 public class FieldManager {
-    
+
     private static final Log LOG = LogFactory.getLog(FieldManager.class);
 
     public static final String MAPFIELDSEPARATOR = "---";
 
     protected EntityConfiguration entityConfiguration;
     protected EntityManager entityManager;
     protected List<SortableValue> middleFields = new ArrayList<SortableValue>(5);
+    protected Set<Class> classes = new HashSet<>();
+    protected Map<Class, Map<String, Field>> classFields = new HashMap<>();
 
     public FieldManager(EntityConfiguration entityConfiguration, EntityManager entityManager) {
         this.entityConfiguration = entityConfiguration;
         this.entityManager = entityManager;
+        if (entityManager != null) {
+            Set<EntityType<?>> managedEntities = entityManager.getMetamodel().getEntities();
+            for (EntityType managedEntity : managedEntities) {
+                classes.add(managedEntity.getJavaType());
+            }
+        }
     }
 
     public static Field getSingleField(Class<?> clazz, String fieldName) throws IllegalStateException {
         return BLCFieldUtils.getSingleField(clazz, fieldName);
     }
 
     public Field getField(Class<?> clazz, String fieldName) throws IllegalStateException {
-        DynamicEntityDao dynamicEntityDao = getPersistenceManager(clazz).getDynamicEntityDao();
-        SessionFactory sessionFactory = dynamicEntityDao.getDynamicDaoHelper().
-                getSessionFactory((HibernateEntityManager) dynamicEntityDao.getStandardEntityManager());
-        BLCFieldUtils fieldUtils = new BLCFieldUtils(sessionFactory, true, dynamicEntityDao.useCache(),
-                dynamicEntityDao.getEjb3ConfigurationDao(), entityConfiguration,
-                dynamicEntityDao.getDynamicDaoHelper());
-        return fieldUtils.getField(clazz, fieldName);
+        Map<String, Field> stringFieldMap = classFields.get(clazz);
+        if(stringFieldMap==null){
+            stringFieldMap = new HashMap<>();
+            classFields.put(clazz, stringFieldMap);
```

Comment:
Do we need to be worried about thread safety here?

Also, I looked over in `BLCFieldUtils` and that already has some caching there for `Field` instances (in `getSingleField()`). Maybe the part that it's missing is the class -> `Field` mapping? Or that the caching is not around `getField()`?

Reply:
I was thinking about concurrency here and what I found is that the entry point was a prototype bean, so for each reqeust it should start new instance of all that stuff and no instances are shared between threads. But that's one of things I was hoping you/Jeff can point out if I miss something in the flows where this piece of code is used.

About BLCFieldUtils, hmm, I missed that it already has a cache but profiler didn't reveal that this cache is really used. I think because for each call of a org.broadleafcommerce.openadmin.server.service.persistence.module.FieldManager#getField new instance of `BLCFieldUtils` is created.
```java
            SessionFactory sessionFactory = dynamicEntityDao.getDynamicDaoHelper().
                    getSessionFactory((HibernateEntityManager) dynamicEntityDao.getStandardEntityManager());
            BLCFieldUtils fieldUtils = new BLCFieldUtils(sessionFactory, true, dynamicEntityDao.useCache(),
                    dynamicEntityDao.getEjb3ConfigurationDao(), entityConfiguration,
                    dynamicEntityDao.getDynamicDaoHelper());
            Field fld = fieldUtils.getField(clazz, fieldName);
```
Probably my cache can be removed and instance of `BLCFieldUtils` is created per `FieldManager` instance, if at `FieldManager` construction time it has all needed properties, which might be a little tricky.

Reply:
checked for null and create one instance of BLCFieldUtils per FieldManager instance

Reply:
> I was thinking about concurrency here and what I found is that the entry point was a prototype bean, so for each reqeust it should start new instance of all that stuff and no instances are shared between threads.

That's true, should be ok.

> About BLCFieldUtils, hmm, I missed that it already has a cache but profiler didn't reveal that this cache is really used. I think because for each call of a org.broadleafcommerce.openadmin.server.service.persistence.module.FieldManager#getField new instance of BLCFieldUtils is created.

Yes, but the `FIELD_CACHE` is static (and usage is in a `synchronized` block) which means that it's shared across all instances. It seems like it is more likely that in the particular flow you were looking at is not actually exercising anything in the `FIELD_CACHE` for whatever reason.

Regardless, if your change here to have a single instance of the `FieldManager` still resolves the perf issues you were seeing then I think you're ok.